data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarder.java

```diff
@@ -0,0 +1,114 @@
+
+public class PeerForwarder {
+    private static final WebClient LOCAL_CLIENT = null;
+
+    private final PeerForwarderClientFactory peerForwarderClientFactory;
+    private final PeerForwarderClient peerForwarderClient;
+    private HashRing hashRing;
+    private PeerClientPool peerClientPool;
+
+    public PeerForwarder(final PeerForwarderClientFactory peerForwarderClientFactory,
+                         final PeerForwarderClient peerForwarderClient) {
+        this.peerForwarderClientFactory = peerForwarderClientFactory;
+        this.peerForwarderClient = peerForwarderClient;
+    }
+
+    public List<Record<Event>> forwardRecords(final Collection<Record<Event>> records, final Set<String> identificationKeys) {
+        // TODO: decide the default behaviour of Core Peer Forwarder and move the HashRing and PeerClientPool creation to constructor
+        hashRing = peerForwarderClientFactory.createHashRing();
+        peerClientPool = peerForwarderClientFactory.setPeerClientPool();
+
+        final Map<String, List<Record<Event>>> groupedRecords = groupRecordsBasedOnIdentificationKeys(records, identificationKeys);
+
+        final List<Record<Event>> recordsToProcessLocally = new ArrayList<>();
+
+        for (final Map.Entry<String, List<Record<Event>>> entry: groupedRecords.entrySet()) {
+            final WebClient client = getClient(entry.getKey());
+
+            if (isLocalClient(client)) {
+                recordsToProcessLocally.addAll(entry.getValue());
+            }
+            else {
+                AggregatedHttpResponse httpResponse = peerForwarderClient.serializeRecordsAndSendHttpRequest(entry.getValue(), client);
+                if (httpResponse == null || httpResponse.status() != HttpStatus.OK) {
+                    recordsToProcessLocally.addAll(entry.getValue());
+                }
+            }
+        }
+        return recordsToProcessLocally;
+    }
+
+    private Map<String, List<Record<Event>>> groupRecordsBasedOnIdentificationKeys(
+            final Collection<Record<Event>> records,
+            final Set<String> identificationKeys
+    ) {
+        final Map<String, List<Record<Event>>> groupedRecords = new HashMap<>();
+
+        // group records based on IP address calculated by HashRing
+        for (final Record<Event> record : records) {
+            final Event event = record.getData();
+
+            List<String> identificationKeyValues = new LinkedList<>();
+            for (final String identificationKey : identificationKeys) {
+                identificationKeyValues.add(event.get(identificationKey, Object.class).toString());
+            }
+
+            final String dataPrepperIp = hashRing.getServerIp(identificationKeyValues).orElse(StaticPeerListProvider.LOCAL_ENDPOINT);
+            groupedRecords.computeIfAbsent(dataPrepperIp, x -> new ArrayList<>()).add(record);
+        }
+        return groupedRecords;
+    }
+
+    private WebClient getClient(final String address) {
```

Comment:
Rather than creating the `WebClient` here and passing it into the `PeerForwarderClient`, I think it would be ideal to pass only the serverIp address to `PeerForwarderClient`. It would be responsible for getting the client from the pool.